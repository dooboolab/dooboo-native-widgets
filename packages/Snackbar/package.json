{
  "name": "@dooboo-ui/snackbar",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "version": "0.0.1",
  "license": "MIT",
  "postinstall": "tsc",
  "scripts": {
    "build": "rm -rf ./lib && tsc",
    "watch": "tsc -w"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/dooboolab/dooboo-ui.git",
    "directory": "src/components/shared/Snackbar"
  },
  "bugs": {
    "url": "https://github.com/dooboolab/dooboo-ui/issues"
  },
  "homepage": "https://github.com/dooboolab/dooboo-ui#readme",
  "dependencies": {
    "dooboolab-welcome": "^1.1.1"
  },
  "devDependencies": {
    "styled-components": "^4.4.0",
    "typescript": "^3.6.4"
  },
  "peerDependencies": {
    "react": ">=16.8",
    "react-native": ">=0.59",
    "styled-components": ">=4.4.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "keywords": [
    "snackbar",
    "react-hook",
    "react-native",
    "dooboo-ui/native",
    "dooboolab"
  ]
}
