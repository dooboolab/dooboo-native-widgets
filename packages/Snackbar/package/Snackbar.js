import * as React from 'react';
import { Animated, Dimensions, StyleSheet } from 'react-native';
import styled from 'styled-components/native';
var __makeTemplateObject = (this && this.__makeTemplateObject) || function(cooked, raw) {
  if (Object.defineProperty) { Object.defineProperty(cooked, 'raw', { value: raw }); } else { cooked.raw = raw; }
  return cooked;
};
var width = Dimensions.get('screen').width;
var maxWidth = width - 32;
var styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
    minWidth: 150,
    maxWidth: maxWidth,
    textAlign: 'left',
    alignItems: 'center',
    alignSelf: 'center',
    position: 'absolute',
    fontSize: 16,
    paddingHorizontal: 16,
    paddingVertical: 10,
    bottom: 10,
    backgroundColor: '#87b5ff',
    borderRadius: 10,
  },
});
var MessageText = styled.Text(templateObject_1 || (templateObject_1 = __makeTemplateObject(['\n  color: white;\n'], ['\n  color: white;\n'])));
var ActionText = styled.Text(templateObject_2 || (templateObject_2 = __makeTemplateObject(['\n  color: green;\n'], ['\n  color: green;\n'])));
var ActionContainer = styled.View(templateObject_3 || (templateObject_3 = __makeTemplateObject(['\n  display: flex;\n  align-items: center;\n  margin-left: auto;\n  margin-right: -5px;\n  padding-left: 16px;\n'], ['\n  display: flex;\n  align-items: center;\n  margin-left: auto;\n  margin-right: -5px;\n  padding-left: 16px;\n'])));
var Touchable = styled.TouchableOpacity(templateObject_4 || (templateObject_4 = __makeTemplateObject([''], [''])));
var ActionButton = styled.View(templateObject_5 || (templateObject_5 = __makeTemplateObject(['\n  padding: 4px 4px 2px 2px;\n'], ['\n  padding: 4px 4px 2px 2px;\n'])));
export var Timer;
(function(Timer) {
  Timer[Timer.SHORT = 1500] = 'SHORT';
  Timer[Timer.LONG = 3000] = 'LONG';
})(Timer || (Timer = {}));
var Snackbar = function(props, ref) {
  var testID = props.testID;
  var _a = React.useState({ isVisible: false, isShowing: false }); var showingState = _a[0]; var setShowingState = _a[1];
  var _b = React.useState({ text: '', timer: Timer.SHORT }); var content = _b[0]; var setContent = _b[1];
  var text = content.text; var actionText = content.actionText; var messageStyle = content.messageStyle; var actionStyle = content.actionStyle; var containerStyle = content.containerStyle; var _c = content.timer; var timer = _c === void 0 ? Timer.SHORT : _c; var onPressAction = content.onPressAction;
  var isShowing = showingState.isShowing; var isVisible = showingState.isVisible; var timeout = showingState.timeout;
  var fadeAnim = React.useState(new Animated.Value(0))[0];
  var show = function(c) {
    setContent(c);
    timeout && clearTimeout(timeout);
    setShowingState(function(prevState) { return Object.assign(Object.assign({}, prevState), { isShowing: true }); });
  };
  var hide = function(duration) {
    if (duration === void 0) { duration = 200; }
    Animated.timing(fadeAnim, {
      toValue: 0,
      duration: duration,
      useNativeDriver: true,
    }).start(function() { return setShowingState(function(prevState) { return Object.assign(Object.assign({}, prevState), { isVisible: false }); }); });
  };
  React.useEffect(function() {
    if (isShowing) {
      if (isVisible) {
        hide(50);
      } else {
        var hideTimeout = setTimeout(function() {
          hide();
        }, timer + 200);
        setShowingState({ isShowing: false, isVisible: true, timeout: hideTimeout });
        Animated.timing(fadeAnim, {
          toValue: 1,
          duration: 200,
          useNativeDriver: true,
        }).start();
      }
    }
  }, [showingState]);
  React.useImperativeHandle(ref, function() {
    return ({
      show: show,
    });
  });
  return (<>
    {showingState.isVisible && (<Animated.View testID={testID} style={[styles.container, containerStyle, { opacity: fadeAnim }]}>
      <MessageText style={messageStyle}>{text}</MessageText>
      {actionText && (<ActionContainer>
        <Touchable onPress={onPressAction}>
          <ActionButton>
            <ActionText style={actionStyle}>{actionText}</ActionText>
          </ActionButton>
        </Touchable>
      </ActionContainer>)}
    </Animated.View>)}
  </>);
};
export default React.forwardRef(Snackbar);
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5;
