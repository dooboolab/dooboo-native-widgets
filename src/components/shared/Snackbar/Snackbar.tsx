import * as React from 'react';

import {
  Animated, Dimensions, StyleSheet, TextStyle, ViewStyle,
} from 'react-native';

import styled from 'styled-components/native';

const { width } = Dimensions.get('screen');
const maxWidth = width - 32;

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
    minWidth: 150,
    maxWidth,
    textAlign: 'left',
    alignItems: 'center',
    alignSelf: 'center',
    position: 'absolute',
    fontSize: 16,
    paddingHorizontal: 16,
    paddingVertical: 10,
    bottom: 10,
    backgroundColor: '#87b5ff',
    borderRadius: 10,
  },
});

const MessageText = styled.Text`
  color: white;
`;

const ActionText = styled.Text`
  color: green;
`;

const ActionContainer = styled.View`
  display: flex;
  align-items: center;
  margin-left: auto;
  margin-right: -5px;
  padding-left: 16px;
`;

const Touchable = styled.TouchableOpacity``;

const ActionButton = styled.View`
  padding: 4px 4px 2px 2px;
`;

export interface SnackbarProps {
  testID?: string;
  ref: React.MutableRefObject<SnackbarRef>;
}

export interface Content {
  text: string;
  actionText?: string;
  timer?: Timer;
  actionStyle?: TextStyle;
  containerStyle?: ViewStyle;
  messageStyle?: TextStyle;
  onPressAction?: () => void;
}

interface ShowingState {
  isVisible?: boolean;
  isShowing?: boolean;
  timeout?: NodeJS.Timeout;
}

export interface SnackbarRef {
  show(content: Content): void;
}

export enum Timer {
  SHORT = 1500,
  LONG = 3000,
}

const Snackbar = (props: SnackbarProps, ref: React.Ref<SnackbarRef>): React.ReactElement => {
  const { testID } = props;
  const [showingState, setShowingState] = React.useState<ShowingState>(
    { isVisible: false, isShowing: false },
  );
  const [content, setContent] = React.useState<Content>({ text: '', timer: Timer.SHORT });
  const {
    text, actionText, messageStyle, actionStyle, containerStyle, timer = Timer.SHORT, onPressAction,
  } = content;
  const { isShowing, isVisible, timeout } = showingState;
  const [fadeAnim] = React.useState(new Animated.Value(0));
  const show = (c: Content): void => {
    setContent(c);
    timeout && clearTimeout(timeout);
    setShowingState((prevState) => Object.assign(Object.assign({}, prevState), { isShowing: true }));
  };
  const hide = (duration = 200): void => {
    Animated.timing(
      fadeAnim,
      {
        toValue: 0,
        duration,
      },
    ).start(() => setShowingState(
      (prevState) => Object.assign(Object.assign({}, prevState), { isVisible: false }),
    ));
  };
  React.useEffect(() => {
    if (isShowing) {
      if (isVisible) {
        hide(50);
      } else {
        const hideTimeout = setTimeout(() => {
          hide();
        }, timer + 200);
        setShowingState({ isShowing: false, isVisible: true, timeout: hideTimeout });
        Animated.timing(
          fadeAnim,
          {
            toValue: 1,
            duration: 200,
          },
        ).start();
      }
    }
  }, [showingState]);
  React.useImperativeHandle(ref, () => ({
    show,
  }));
  return (
    <>
      {showingState.isVisible && (
        <Animated.View
          testID={testID}
          style={[styles.container, containerStyle, { opacity: fadeAnim }]}
        >
          <MessageText style={messageStyle}>{text}</MessageText>
          {actionText && (
            <ActionContainer>
              <Touchable onPress={onPressAction}>
                <ActionButton>
                  <ActionText style={actionStyle}>{actionText}</ActionText>
                </ActionButton>
              </Touchable>
            </ActionContainer>
          )}
        </Animated.View>
      )}
    </>
  );
};

export default React.forwardRef<SnackbarRef, SnackbarProps>(Snackbar);
