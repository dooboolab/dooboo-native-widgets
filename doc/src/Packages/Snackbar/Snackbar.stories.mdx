import { Meta, Story, Canvas } from '@storybook/addon-docs/blocks';
import { Simple, StyleChanged, TextStyleChanged } from './examples';

<Meta title="Packages/Snackbar" />

# Snackbar
Simple snackbar component. Snackbar shows the temporal message at bottom of the screen.
If you want to see the way of installation and basic usage, see [here](https://github.com/dooboolab/dooboo-ui/tree/master/packages/Snackbar).

## Examples

Snackbar is a component but it needs different ways to use from other components. To shows Snackbar, the snackbar should be added to whole screen view and call **'show'** method of the refrence.

|                | necessary | types            | default                      | info               |
| -------------- | :-------: | ---------------- | ---------------------------- | ------------------ |
| text           |     v     | string           | undefined                    |                    |
| timer          |           | number           | 1500 (Timer.SHORT)           | Closing time (ms)  |
| messageStyle   |           | TextStyle        | { color: 'white' }           | Message text style |
| containerStyle |           | ViewStyle        | [See here](#containerstyle)  |                    |
| actionText     |           | string           | undefined                    |                    |
| actionStyle    |           | TextStyle        | { color: 'green' }           | Action text style  |
| onPressAction  |           | function         | undefined                    |                    |


### Message Contents and duration

You can show a Snackbar with messages during some milliseconds like this.
```typescript
snackbar.show({
  text: 'Simple text',
  actionText: 'Some action text',
  timer: 3000,
})
```
And you can see how Snackbar actually appears by changing the parameter values in the block below.

<Canvas withSource='none'>
    <Simple />
</Canvas>

Set the action text to show the touchable.

### View style

The default style of snackbar container is like below.
```typescript
{
  display: 'flex',
  flexDirection: 'row',
  flexWrap: 'wrap',
  minWidth: 150,
  maxWidth: Dimensions.get('screen') - 32,
  textAlign: 'left',
  alignItems: 'center',
  alignSelf: 'center',
  position: 'absolute',
  fontSize: 16,
  paddingHorizontal: 16,
  paddingVertical: 10,
  bottom: 10,
  backgroundColor: '#87b5ff',
  borderRadius: 10,
}
```

You can change the container style by adding `containerStyle` field. 
```typescript
snackbar.show({
  text: 'Simple text message',
  containerStyle: {
    alignSelf: 'left',
    backgroundColor: '#0008',
    bottom: undefined, // if this line removed then the component will be stretched vertically because of default bottom value (10).
    up: 10,
    minWidth: 300,
  }
})
```
<Canvas withSource='none'>
  <StyleChanged />
</Canvas>

### Text styles
You can also change text styles by add `messageStyle` and `actionStyle` fields to the `show` method parameter like below.
```typescript
snackbar.show({
  text: 'Simple text message',
  messageStyle: {
    color: '#ff0000',
    fontSize: 14,
  },
  actionText: 'Some Action',
  actionStyle: {
    fontSize: 22,
  },
})
```
<Canvas withSource='none'>
  <TextStyleChanged />
</Canvas>
