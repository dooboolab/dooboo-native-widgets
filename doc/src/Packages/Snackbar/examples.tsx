import React, { ReactElement, useCallback, useRef, useState } from 'react';
import Snackbar, {
  SnackbarProvider,
  SnackbarRef,
  Timer,
  useSnackbarContext,
  Content,
} from '@dooboo-ui/snackbar';

import styled from 'styled-components/native';
import { Button, EditText } from 'dooboo-ui';

const Container = styled.View`
  background-color: transparent;
  align-items: center;
  flex: 1;
  min-height: 400px;
`;

export function Simple(): React.ReactElement {
  const snackbar = useRef<SnackbarRef>();
  const [ text, setText ] = useState<string>('Simple text');
  const [ actionText, setActionText ] = useState<string>('');
  const [ timer, setTimer ] = useState('1500');
  const onPress = () => {
    snackbar?.current?.show({
      text,
      timer: +timer,
      actionText,
    })
  }
  return <Container>
    <EditText label="text" onChangeText={(changedText) => setText(changedText)} value={text}/>
    <EditText label="timer" onChangeText={(timer) => setTimer(timer)} value={timer}/>
    <EditText label="actionText" onChangeText={(changedText) => setActionText(changedText)} value={actionText}/>
    <Button text="Click this to show snackbar" onPress={onPress} containerStyle={{ backgroundColor: "#eee", borderRadius: 4, margin: 10 }} />
    <Snackbar ref={snackbar} />
  </Container>
}

function getExample(param: Partial<Content>) {
  return (): React.ReactElement => {
    const snackbar = useRef<SnackbarRef>();
    const onPress = () => {
      snackbar?.current?.show({
        text: 'Simple text message',
        ...param
      });
    }
    return <Container style={{justifyContent: 'center'}}>
      <Button text="Click this to show snackbar" onPress={onPress} containerStyle={{ backgroundColor: "#eee", borderRadius: 4, margin: 10 }} />
      <Snackbar ref={snackbar} />
    </Container>
  }
}

export const StyleChanged = getExample({
  containerStyle: {
    alignSelf: 'flex-start',
    backgroundColor: '#0008',
    bottom: undefined,
    top: 10,
    minWidth: 300,
  }      
});

export const TextStyleChanged = getExample({
  messageStyle: {
    color: '#ff0000',
    fontSize: 14,
  },
  actionText: 'Some Action',
  actionStyle: {
    fontSize: 22,
  }
});