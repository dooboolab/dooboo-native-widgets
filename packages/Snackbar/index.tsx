import React, { createContext, useContext, useRef } from 'react';
import Snackbar, { Content, SnackbarRef } from './Snackbar';
import { View } from 'react-native';

interface SnackbarContext {
  show(content: Content): void;
}

const SnackbarContext = createContext<SnackbarContext | undefined>(undefined);
const useCtx = (): SnackbarContext | undefined => {
  const c = useContext<SnackbarContext | undefined>(SnackbarContext);
  if (!c) {
    throw new Error('useCtx must be inside a Provider with a value');
  }
  return c;
};

export interface SnackbarProviderProps {
  children?: React.ReactElement;
  defaultContent?: Partial<Content>;
}

function SnackbarProvider(props: SnackbarProviderProps): React.ReactElement {
  const { children, defaultContent = {} } = props;
  const snackbar = useRef<SnackbarRef>() as React.MutableRefObject<SnackbarRef>;
  const show = (content: Content): void => {
    snackbar.current && snackbar.current.show({ ...defaultContent, ...content });
  };
  return (
    <View style={{ flex: 1, width: '100%' }}>
      <SnackbarContext.Provider value={{ show }}>{children}</SnackbarContext.Provider>
      <Snackbar ref={snackbar} />
    </View>);
}

export { useCtx as useSnackbarContext, SnackbarProvider };
export * from './Snackbar';
export default Snackbar;
