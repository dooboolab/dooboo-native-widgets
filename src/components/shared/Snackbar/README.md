# Snackbar

[![Npm Version](http://img.shields.io/npm/v/@dooboo-ui/native-snackbar.svg?style=flat-square)](https://npmjs.org/package/@dooboo-ui/native-snackbar)
[![Downloads](http://img.shields.io/npm/dm/@dooboo-ui/native-snackbar.svg?style=flat-square)](https://npmjs.org/package/@dooboo-ui/native-snackbar)

> Simple snackbar for react-native.

![Jan-28-2020 15-59-38](https://user-images.githubusercontent.com/17980230/73242107-548d0400-41e7-11ea-946b-630ba7053584.gif)

## Installation

At this point, this component has not yet been published, and after it has been published, it may be installed with the command below.

```sh
yarn add @dooboo-ui/native
```

or

```sh
yarn add @dooboo-ui/native-snackbar
```

## Usage

# Types

```ts
export enum Timer {
  SHORT = 1500,
  LONG = 3000,
}

export interface SnackbarProps {
  testID?: string;
  ref: React.MutableRefObject<SnackbarRef>;
}

export interface SnackbarRef {
  show(content: Content): void;
}

export interface Content {
  text: string;
  actionText?: string;
  timer?: Timer;
  actionStyle?: TextStyle;
  containerStyle?: ViewStyle;
  messageStyle?: TextStyle;
  onPressAction?: () => void;
}
```

- SnackbarProps

|         | necessary | types                   | default   | info         |
| ------- | :-------: | ----------------------- | --------- | ------------ |
| ref     |     v     | MutableRefObject        | undefined |              |
| testID  |           | string                  | undefined |              |

- Content

An object of this type is needed to show an Snackbar. 

|                | necessary | types            | default            | info               |
| -------------- | :-------: | ---------------- | ------------------ | ------------------ |
| text           |     v     | string           | undefined          |                    |
| timer          |           | number           | 1500 (Timer.SHORT) | Closing time       |
| messageStyle   |           | TextStyle        | undefined          | Message text style |
| containerStyle |           | ViewStyle        | undefined          |                    |
| actionText     |           | string           | undefined          |                    |
| actionStyle    |           | TextStyle        | undefined          | Action text style  |
| onPressAction  |           | function         | undefined          |                    |

# Getting started

- Import

  ```tsx
  import { Snackbar } from '@dooboo-ui/native';
  // or
  import Snackbar from '@dooboo-ui/native-snackbar';
  ```

- Usage
  ```tsx

  function Container(): React.ReactElement {
    const snackbar = useRef<SnackbarRef>();
    const onPress = (): void => {
      snackbar.current && snackbar.current.show({
        text: 'Simple Snackbar is opened',
        timer: Timer.LONG,
        containerStyle: {
          backgroundColor: '#ccccff',
        },
        messageStyle: {
          color: '#ffffff',
          fontSize: 17,
        },
      actionText: 'ACTION',
      actionStyle: {
        color: '#0066ff',
        fontSize: 17,
      },
      onPressAction: () => Alert.alert('Action!!'),
      });
    };

    return (
      <Container>
        <Button onPress={onPress}>
          <Text style={{ textAlign: 'center' }}>OPEN SNACKBAR</Text>
        </Button>
        <Snackbar ref={snackbar}/>
      </Container>
    );
  }

  ```
To show a `Snackbar` component, just provide `ref` props to the component and call the `show` function  (with a `Content` type parameter) of it.
This component will appear at the bottom of the parent view, **not at the bottom of the screen**.
