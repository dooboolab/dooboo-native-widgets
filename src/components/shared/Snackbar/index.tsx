import React, { createContext, useContext, useRef } from 'react';
import { SafeAreaView, View } from 'react-native';
import Snackbar, { Content, SnackbarRef } from './Snackbar';

interface SnackbarContext {
  show(content: Content): void;
}

const SnackbarContext = createContext<SnackbarContext | undefined>(undefined);
const useCtx = (): SnackbarContext | undefined => {
  const c = useContext<SnackbarContext | undefined>(SnackbarContext);
  if (!c) {
    throw new Error('useCtx must be inside a Provider with a value');
  }
  return c;
};

export interface SnackbarProviderProps {
  children?: React.ReactElement;
  useWholeScreen?: boolean;
}

function SnackbarProvider(props: SnackbarProviderProps): React.ReactElement {
  const { children, useWholeScreen } = props;
  const snackbar = useRef<SnackbarRef>() as React.MutableRefObject<SnackbarRef>;
  const show = (content: Content): void => {
    snackbar.current && snackbar.current.show(content);
  };
  const Container = useWholeScreen ? View : SafeAreaView;
  return (
    <Container style={{ flex: 1, alignItems: 'center' }}>
      <SnackbarContext.Provider value={{ show }}>{children}</SnackbarContext.Provider>
      <Snackbar ref={snackbar} />
    </Container>);
}

export { useCtx as useSnackbarContext, SnackbarProvider };
export * from './Snackbar';
export default Snackbar;
