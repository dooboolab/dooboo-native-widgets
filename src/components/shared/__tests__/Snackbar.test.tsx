import * as React from 'react';

import { Animated, Text, TouchableOpacity, View } from 'react-native';
import Snackbar, { SnackbarProvider, SnackbarRef, useSnackbarContext } from '../Snackbar';

import { fireEvent, render, wait } from '@testing-library/react-native';
import renderer, { act } from 'react-test-renderer';

// Note: test renderer must be required after react-native.
let snackbarRef: React.MutableRefObject<SnackbarRef>;

jest.useFakeTimers();

const buttonText = 'show snack bar';
const message = 'snackbar content';

function TestWrapper(): React.ReactElement {
  snackbarRef = React.useRef();
  return (
    <View style={{ flex: 1 }}>
      <TouchableOpacity
        testID="Button"
        onPress={(): void => {
          snackbarRef.current && snackbarRef.current.show({
            text: message,
            actionText: 'some action',
          });
        }}>
        <Text>
          {buttonText}
        </Text>
      </TouchableOpacity>
      <Snackbar testID={'snackbar'} ref={snackbarRef} />
    </View>
  );
}

function ContentInnerProvider(): React.ReactElement {
  const snackbar = useSnackbarContext();
  return (
    <View style={{ flex: 1 }}>
      <TouchableOpacity
        testID="Button"
        onPress={(): void => {
          snackbar.show({
            text: message,
            actionText: 'some action',
          });
        }}>
        <Text>
          {buttonText}
        </Text>
      </TouchableOpacity>
    </View>
  );
}

describe('[Snackbar]', () => {
  it('renders TestWrapper without crashing', () => {
    const rendered = renderer.create(<TestWrapper />);
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it('should simulate showing snackbar', async () => {
    const renderResult = render(<TestWrapper />);
    const btn = renderResult.getByTestId('Button');
    act(() => {
      fireEvent.press(btn);
    });
    await wait(() => expect(renderResult.getByTestId('snackbar')).toBeTruthy());
    expect(renderResult.asJSON()).toMatchSnapshot();
    act(() => {
      jest.runAllTimers();
    });
    expect(renderResult.queryByTestId('snackbar')).toBeFalsy();

    // Test hide previous snackbar when showing new one.
    const timingSpy = jest.spyOn(Animated, 'timing');

    act(() => {
      fireEvent.press(btn);
    });
    await wait(() => expect(renderResult.getByTestId('snackbar')).toBeTruthy());
    act(() => {
      fireEvent.press(btn);
    });

    // Check if close(50) called!
    expect(timingSpy.mock.calls.find(
      (predicate) => predicate[1].toValue === 0 && predicate[1].duration === 50),
    ).toBeTruthy();

    act(() => jest.runAllTimers());

    // Check hide
    expect(renderResult.queryByTestId('snackbar')).toBeFalsy();
  });
});

describe('[Snackbar] using provider', () => {
  const TestElement = (useWholeScreen = false): React.ReactElement => (
    <SnackbarProvider useWholeScreen={useWholeScreen}>
      <ContentInnerProvider/>
    </SnackbarProvider>
  );

  it('renders TestElement without crashing', () => {
    const rendered = renderer.create(TestElement());
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it('renders TestElement using whole screen view without crashing', () => {
    const rendered = renderer.create(TestElement(true));
    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it('occurs error when use snackbar context outside of the SnackbarProvider', async () => {
    expect(() => renderer.create(<ContentInnerProvider/>)).toThrowError();
  });

  it('should simulate showing snackbar', async () => {
    const renderResult = render(TestElement());
    const btn = renderResult.getByTestId('Button');
    act(() => {
      fireEvent.press(btn);
    });
    await wait(() => expect(renderResult.getByText(message)).toBeTruthy());
    expect(renderResult.asJSON()).toMatchSnapshot();
    act(() => {
      jest.runAllTimers();
    });
    expect(renderResult.queryByTestId('snackbar')).toBeFalsy();

    // Test hide previous snackbar when showing new one.
    const timingSpy = jest.spyOn(Animated, 'timing');

    act(() => {
      fireEvent.press(btn);
    });
    await wait(() => expect(renderResult.getByText(message)).toBeTruthy());
    act(() => {
      fireEvent.press(btn);
    });

    // Check if close(50) called!
    expect(timingSpy.mock.calls.find(
      (predicate) => predicate[1].toValue === 0 && predicate[1].duration === 50),
    ).toBeTruthy();

    act(() => jest.runAllTimers());

    // Check hide
    expect(renderResult.queryByTestId('snackbar')).toBeFalsy();
  });
});
