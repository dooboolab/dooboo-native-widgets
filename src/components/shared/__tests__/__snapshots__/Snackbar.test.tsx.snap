// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Snackbar] renders TestWrapper without crashing 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="Button"
  >
    <Text>
      show snack bar
    </Text>
  </View>
</View>
`;

exports[`[Snackbar] should simulate showing snackbar 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="Button"
    >
      <Text>
        show snack bar
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#303235",
          "borderRadius": 10,
          "bottom": 10,
          "display": "flex",
          "flexDirection": "row",
          "flexWrap": "wrap",
          "fontSize": 16,
          "maxWidth": 718,
          "minWidth": 150,
          "opacity": 0,
          "paddingHorizontal": 16,
          "paddingVertical": 10,
          "position": "absolute",
          "textAlign": "left",
        }
      }
      testID="snackbar"
    >
      <Text>
        snackbar content
      </Text>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "display": "flex",
              "marginLeft": "auto",
              "marginRight": -5,
              "paddingLeft": 16,
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={false}
          isTVSelectable={true}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "paddingBottom": 2,
                  "paddingLeft": 2,
                  "paddingRight": 4,
                  "paddingTop": 4,
                },
              ]
            }
          >
            <Text>
              some action
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
